# Linear Regression with nn.Module in PyTorch

In [1]:
# Import Module
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(1)

In [2]:
x_train = torch.FloatTensor([[1],[2],[3]])
y_train = torch.FloatTensor([[2],[4],[6]])

In [3]:
# input_dim = 1, output_dim=1
model = nn.Linear(1,1)

In [4]:
print(list(model.parameters()))

[Parameter containing:
tensor([[0.5153]], requires_grad=True), Parameter containing:
tensor([-0.4414], requires_grad=True)]


In [5]:
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

In [6]:
nb_epochs = 2001
for epoch in range(nb_epochs):
    prediction = model(x_train)
    cost = F.mse_loss(prediction, y_train)

    optimizer.zero_grad()
    cost.backward()
    optimizer.step()

    if epoch%100==0:
        print('Epoch {:4d}/{} Cost: {:.6f}'.format(epoch, nb_epochs, cost.item()))

Epoch    0/2001 Cost: 13.103541
Epoch  100/2001 Cost: 0.002791
Epoch  200/2001 Cost: 0.001724
Epoch  300/2001 Cost: 0.001066
Epoch  400/2001 Cost: 0.000658
Epoch  500/2001 Cost: 0.000407
Epoch  600/2001 Cost: 0.000251
Epoch  700/2001 Cost: 0.000155
Epoch  800/2001 Cost: 0.000096
Epoch  900/2001 Cost: 0.000059
Epoch 1000/2001 Cost: 0.000037
Epoch 1100/2001 Cost: 0.000023
Epoch 1200/2001 Cost: 0.000014
Epoch 1300/2001 Cost: 0.000009
Epoch 1400/2001 Cost: 0.000005
Epoch 1500/2001 Cost: 0.000003
Epoch 1600/2001 Cost: 0.000002
Epoch 1700/2001 Cost: 0.000001
Epoch 1800/2001 Cost: 0.000001
Epoch 1900/2001 Cost: 0.000000
Epoch 2000/2001 Cost: 0.000000


In [7]:
# Test
new_var = torch.FloatTensor([[4.0]])
pred_y = model(new_var)
print("훈련 후 입력이 4일 때의 예측값 : ", pred_y)

훈련 후 입력이 4일 때의 예측값 :  tensor([[7.9989]], grad_fn=<AddmmBackward0>)


해당 문제의 정답은 $$y=2x$$ 이므로 $$y$$값이 8에 가까우면 가중치와 편향이 어느정도 최적화가 되었다고 볼 수 있습니다.

In [8]:
print(list(model.parameters()))

[Parameter containing:
tensor([[1.9994]], requires_grad=True), Parameter containing:
tensor([0.0014], requires_grad=True)]


# Multivariable Linear Regression with nn.Module in PyTorch

In [48]:
# 데이터
xx_train = torch.FloatTensor([[73, 80, 75],
                             [93, 88, 93],
                             [89, 91, 90],
                             [96, 98, 100],
                             [73, 66, 70]])
yy_train = torch.FloatTensor([[152], [185], [180], [196], [142]])

In [49]:
mmodel = nn.Linear(3, 1)

In [50]:
print(list(mmodel.parameters()))

[Parameter containing:
tensor([[-0.0968, -0.2490, -0.1850]], requires_grad=True), Parameter containing:
tensor([0.0276], requires_grad=True)]


In [51]:
optimizer = torch.optim.SGD(mmodel.parameters(), lr=1e-5) 

In [55]:
nb_epochs = 10000
for epoch in range(nb_epochs+1):

    # H(x) 계산
    prediction = mmodel(xx_train)
    # model(x_train)은 model.forward(x_train)와 동일함.

    # cost 계산
    cost = F.mse_loss(prediction, yy_train) # <== 파이토치에서 제공하는 평균 제곱 오차 함수

    # cost로 H(x) 개선하는 부분
    # gradient를 0으로 초기화
    optimizer.zero_grad()
    # 비용 함수를 미분하여 gradient 계산
    cost.backward()
    # W와 b를 업데이트
    optimizer.step()

    if epoch % 100 == 0:
    # 100번마다 로그 출력
      print('Epoch {:4d}/{} Cost: {:.6f}'.format(
          epoch, nb_epochs, cost.item()
      ))

Epoch    0/10000 Cost: 0.256865
Epoch  100/10000 Cost: 0.256211
Epoch  200/10000 Cost: 0.255569
Epoch  300/10000 Cost: 0.254935
Epoch  400/10000 Cost: 0.254335
Epoch  500/10000 Cost: 0.253728
Epoch  600/10000 Cost: 0.253141
Epoch  700/10000 Cost: 0.252554
Epoch  800/10000 Cost: 0.251995
Epoch  900/10000 Cost: 0.251436
Epoch 1000/10000 Cost: 0.250886
Epoch 1100/10000 Cost: 0.250344
Epoch 1200/10000 Cost: 0.249821
Epoch 1300/10000 Cost: 0.249289
Epoch 1400/10000 Cost: 0.248781
Epoch 1500/10000 Cost: 0.248275
Epoch 1600/10000 Cost: 0.247781
Epoch 1700/10000 Cost: 0.247292
Epoch 1800/10000 Cost: 0.246805
Epoch 1900/10000 Cost: 0.246337
Epoch 2000/10000 Cost: 0.245863
Epoch 2100/10000 Cost: 0.245404
Epoch 2200/10000 Cost: 0.244945
Epoch 2300/10000 Cost: 0.244490
Epoch 2400/10000 Cost: 0.244046
Epoch 2500/10000 Cost: 0.243612
Epoch 2600/10000 Cost: 0.243166
Epoch 2700/10000 Cost: 0.242744
Epoch 2800/10000 Cost: 0.242320
Epoch 2900/10000 Cost: 0.241889
Epoch 3000/10000 Cost: 0.241475
Epoch 31

In [56]:
# 임의의 입력 [73, 80, 75]를 선언
new_var =  torch.FloatTensor([[73, 80, 75]]) 
# 입력한 값 [73, 80, 75]에 대해서 예측값 y를 리턴받아서 pred_y에 저장
pred_y = mmodel(new_var) 
print("훈련 후 입력이 73, 80, 75일 때의 예측값 :", pred_y) 

훈련 후 입력이 73, 80, 75일 때의 예측값 : tensor([[151.3319]], grad_fn=<AddmmBackward0>)


사실 3개의 값 73, 80, 75는 훈련 데이터로 사용되었던 값입니다.  
당시 y의 값은 152였는데, 현재 예측값이 151이 나온 것으로 보아 어느정도는 3개의 w와 b의 값이 최적화 된것으로 보입니다. 이제 학습 후의 3개의 w와 b의 값을 출력해보겠습니다.

In [57]:
print(list(model.parameters()))

[Parameter containing:
tensor([[0.8541, 0.8475, 0.3096]], requires_grad=True), Parameter containing:
tensor([0.3568], requires_grad=True)]
